### DISCUSSION TITLE
Braking Wheels individually

### DISCUSSION THREAD

**Discussion by aufkeinenfallvergessen** [carla-simulator/carla#5215] [Q&A]
Hello Carla Community,
I want to brake the Vehicle wheels individually to be able to make complex movements like avoiding a collision with another vehicle with a big speed difference. ([as seen in video](https://www.youtube.com/watch?v=zNSnPy-mnhs))

### Always max braking + adjusting max_brake_torque each timestep (always 0 Nm when car shouldnt brake)
Unfortunately the car alway freezes when the max_brake_torque is beeing adjusted ([as seen in video](https://youtu.be/K5YD9IHPbcA)).
I did try it also with "carla.WheelPhysicsControl()" , with stopping the Simualtion and Adjusting the Wheel position with the same outcome.
Is there a way to make it work with that approach without a freezing vehicle?

**ADJUST CONTROL VALUES**
**Get Control Values**
vehicleControlValues = self.vehicle.get_control()
**Update Vehicle Control Variable Values**
vehicleControlValues.brake = 1.0
vehicleControlValues.hand_brake = False
vehicleControlValues.manual_gear_shift = False
vehicleControlValues.reverse = False
vehicleControlValues.steer = 0.0
vehicleControlValues.throttle = 0.5
**Apply Vehicle Control Variable Values**
self.vehicle.apply_control(vehicleControlValues)

**ADJUSTING MAX BRAKE TORQUE**
**Get Wheel Physics**
physics_control = self.vehicle.get_physics_control()
**Update Wheels Physics Control**
wheel_frontLeft = physics_control.wheels[0]
wheel_frontLeft.damping_rate = 700
wheel_frontRight = physics_control.wheels[1]
wheel_frontRight.max_brake_torque = 700
wheel_rearLeft = physics_control.wheels[2]
wheel_rearLeft.max_brake_torque = 700
wheel_rearRight = physics_control.wheels[3]
wheel_rearRight.max_brake_torque = 700
wheels = [wheel_frontRight, wheel_rearLeft, wheel_rearRight, wheel_frontLeft]
**Apply updated Wheels Physics Control**
physics_control.wheels = wheels
self.vehicle.apply_physics_control(physics_control)

### Using the Unreal Engine Function "[USimpleWheeledVehicleMovementComponent](https://docs.unrealengine.com/4.27/en-US/API/Plugins/PhysXVehicles/USimpleWheeledVehicleMovementCom-/)".
This one provides the Function "[SetBrakeTorque](https://docs.unrealengine.com/4.27/en-US/API/Plugins/PhysXVehicles/USimpleWheeledVehicleMovementCom-/SetBrakeTorque/)".
Does anyone know whether I can call it somehow from Carla or how I could implement it to brake the vehicle wheels individually?

I have found the Post [#3586](https://github.com/carla-simulator/carla/issues/3586) with a similar Problem. 
The solution might have been to add those Functions from "[USimpleWheeledVehicleMovementComponent](https://docs.unrealengine.com/4.27/en-US/API/Plugins/PhysXVehicles/USimpleWheeledVehicleMovementCom-/)" to the PythonAPI, as seen in Post [#3868](https://github.com/carla-simulator/carla/pull/3868).
I do not know how to add the Function "[SetBrakeTorque](https://docs.unrealengine.com/4.27/en-US/API/Plugins/PhysXVehicles/USimpleWheeledVehicleMovementCom-/SetBrakeTorque/)" to the Python API myself, since I am not beeing a Carla Developer, but only user.
I use Carla as Debian Package and did build my custom maps in a Docker Container with UE4.
Am i required to build Carla from Source Code to do so?

Cheers :)

**Comment by frankeng** [comment]
Hi 
did you already solve this issue?
I want to brake a vehicle slowly but the minimum deceleration is always -6m/s2 at the beginning, even if I use control.brake =0
![image](https://user-images.githubusercontent.com/23728090/230587080-5ce18454-abcb-4769-a866-79b139eb092c.png)
I tried to do this but it did not help
```
control = local_planner.run_step(debug=False)
physics_control = carla.VehiclePhysicsControl()
physics_control.clutch_strength = 0
physics_control.drag_coefficient = 0
control.throttle = 0
control.bake = 0
actor.apply_physics_control(physics_control)      
actor.apply_control(control)
```
